ddclient

```diff
@@ -2297,29 +2267,36 @@ sub get_ip {
 }
 
 ######################################################################
-## is_ipv4() validates if string is valid IPv4 address and only
-## a valid IPv4 address (no preceding or trailing spaces/characters)
+## is_ipv4() validates if string is valid IPv4 address and only a
+## valid address with no preceding or trailing spaces/characters
+## and no embedded leading zeros.
 ######################################################################
 sub is_ipv4 {
     my ($value) = @_;
     return (length($value // '') != 0) && ($value eq (extract_ipv4($value) // ''));
 }
 
 ######################################################################
-## extract_ipv4() extracts the first valid IPv4 address from the given string
+## extract_ipv4() extracts the first valid IPv4 address from given string.
+## Accepts leading zeros in the address but removes them in returned value
 ######################################################################
 sub extract_ipv4 {
-    (shift // '') =~ /\b((?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?))\b/ai;
-    return $1;
+    if ((shift // '') !~ /\b((?:(?<octet>25[0-5]|2[0-4]\d|[01]?\d\d?)\.){3}
+                         (?&octet))\b/xa) {
+        return undef;
+    }
+    (my $ip = $1) =~ s/\b0+\B//g; ## remove embedded leading zeros
+    return $ip;
 }
 
 ######################################################################
-## is_ipv6() validates if string is valid IPv6 address
+## is_ipv6() validates if string is valid IPv6 address with no preceding
+## or trailing spaces/characters. Embedded leading zeros are accepted
+## but removed for comparison, so extract_ipv6 must remove them also.
 ######################################################################
 sub is_ipv6 {
     my ($value) = @_;
-    # This little gem from http://home.deds.nl/~aeron/regex/
-    return $value =~ /^(((?=.*(::))(?!.*\3.+\3))\3?|([\dA-F]{1,4}(\3|:\b|$)|\2))(?4){5}((?4){2}|(((2[0-4]|1\d|[1-9])?\d|25[0-5])\.?\b){4})\z/ai;
+    return (length($value // '') != 0) && ((extract_ipv6($value) // '') eq (($value =~ s/\b0+\B//g)?$value:$value));
```

Comment:
Interesting approach, and a clever way around the lack of `/r`. A concern I have is whether its understandable by a newcomer examining the code for the first time.

The following are other ways around `/r`, but I don't think they're any more readable:

```perl
do { $value =~ s/\b0+\B//g; $value }
scalar($value =~ s/\b0+\B//g, $value)
```

The following requires more lines of code, but maybe the readability trade-off is worthwhile?

```perl
sub _extract_ipv6 {
    # copy of current extract_ipv6
}

sub extract_ipv6 {
    my $ip = _extract_ipv6(shift) or return undef;
    $ip =~ s/\b0+\B//g;
    return $ip;
}

sub is_ipv6 {
  my ($value) = @_;
  return ($value // '') ne '' && $value eq (_extract_ipv6($value) // '');
}
```

The regular expression version could look like this:

```perl
my $ipv6_regex = qr/.../;

sub extract_ipv6 {
    (shift // '') =~ /\b($ipv6_regex)\b/ or return undef;
    (my $ip = $1) =~ s/\b0+\B//g;
    return $ip;
}

sub is_ipv6 {
    return (shift // '') =~ /^$ipv6_regex$/;
}
```

Reply:
I actually like your last one.... move the regex into a global variable and then use it in both functions.

Reply:
How about we go with what you've got for now, and in PR #198 you can switch to the last one?

Reply:
> How about we go with what you've got for now, and in PR #198 you can switch to the last one?

Yes, because that last one is going to take some testing.